File: desktop/electron/sync-engine.js
/**
 * SyncEngine — watches a local folder and mirrors it to Magizhchi Box cloud.
 *
 * Upload rules:
 *  - New file              → upload to matching cloud folder
 *  - Modified file         → re-upload (server overwrites / creates new version)
 *  - New folder            → create in cloud
 *  - Deleted file/folder   → skip auto-delete (safety first; user can delete via web/mobile)
 *
 * Change detection (avoids re-uploading on restart):
 *  - Tracks { mtime, size } per relative path in a sidecar JSON file.
 *  - File is only uploaded when mtime or size differs from the stored state.
 */

const chokidar = require('chokidar')
const fs = require('fs')
const path = require('path')
const FormData = require('form-data')
const axios = require('axios')

class SyncEngine {
  constructor({ syncFolder, token, deviceId, apiBase, stateFile, onStatus }) {
    this.syncFolder = syncFolder
    this.token = token
    this.deviceId = deviceId
    this.apiBase = apiBase.replace(/\/$/, '') // strip trailing slash
    this.stateFile = stateFile
    this.onStatus = onStatus || (() => {})

    // In-memory caches
    this.folderIdCache = {}   // relFolderPath  → cloudFolderId
    this.syncState    = this._loadState() // relFilePath → { mtime, size }

    // Upload queue (simple FIFO, max 2 concurrent)
    this.queue = []
    this.active = 0
    this.MAX_CONCURRENT = 2

    this.stats = { synced: 0, pending: 0, errors: 0, lastSync: null }
    this.debounceTimers = {}
    this.watcher = null
    this.paused = false
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  getStats() { return { ...this.stats } }

  async start() {
    this._status('syncing', 'Starting initial scan…')
    await this._initialScan()

    this.watcher = chokidar.watch(this.syncFolder, {
      ignored: /(^|[/\\])\.|Thumbs\.db|desktop\.ini/,
      persistent: true,
      ignoreInitial: true,
      awaitWriteFinish: { stabilityThreshold: 1500, pollInterval: 200 },
    })

    this.watcher
      .on('add',    (p) => this._debounce(p, () => this._enqueue('upload', p)))
      .on('change', (p) => this._debounce(p, () => this._enqueue('upload', p)))
      .on('addDir', (p) => { if (p !== this.syncFolder) this._enqueue('mkdir', p) })
      // unlink / unlinkDir → intentionally not auto-deleted on cloud (safety)

    this._status('idle', `Watching · ${this.stats.synced} file(s) synced`)
  }

  stop() {
    this.watcher?.close()
    this.watcher = null
  }

  pause() {
    this.paused = true
    this._status('paused', 'Sync paused')
  }

  resume() {
    this.paused = false
    this._status('idle', 'Resuming sync…')
    this._drain()
  }

  // ── Internal ───────────────────────────────────────────────────────────────

  _debounce(filePath, fn) {
    clearTimeout(this.debounceTimers[filePath])
    this.debounceTimers[filePath] = setTimeout(fn, 800)
  }

  _enqueue(action, filePath) {
    this.queue.push({ action, filePath })
    this.stats.pending = this.queue.length
    this._drain()
  }

  _drain() {
    if (this.paused) return
    while (this.active < this.MAX_CONCURRENT && this.queue.length > 0) {
      const job = this.queue.shift()
      this.stats.pending = this.queue.length
      this.active++
      this._runJob(job).finally(() => {
        this.active--
        this._drain()
        if (this.active === 0 && this.queue.length === 0) {
          this._status('idle', `Up to date · ${this.stats.synced} file(s) synced`)
        }
      })
    }
  }

  async _runJob({ action, filePath }) {
    try {
      if (action === 'upload') await this._uploadFile(filePath)
      else if (action === 'mkdir') await this._ensureCloudFolder(this._rel(filePath))
    } catch (err) {
      this.stats.errors++
      const name = path.basename(filePath)
      this._status('error', `Failed: ${name} — ${err.response?.data?.message || err.message}`)
      console.error('[SyncEngine] job failed:', filePath, err.message)
    }
  }

  async _initialScan() {
    const walk = (dir) => {
      let entries
      try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }
      for (const e of entries) {
        if (e.name.startsWith('.') || e.name === 'Thumbs.db' || e.name === 'desktop.ini') continue
        const full = path.join(dir, e.name)
        if (e.isDirectory()) {
          walk(full)
        } else {
          // Only enqueue if file has changed since last sync
          if (this._hasChanged(full)) {
            this._enqueue('upload', full)
          }
        }
      }
    }
    walk(this.syncFolder)
  }

  _hasChanged(filePath) {
    try {
      const stat = fs.statSync(filePath)
      const rel  = this._rel(filePath)
      const prev = this.syncState[rel]
      if (!prev) return true
      return prev.mtime !== stat.mtimeMs || prev.size !== stat.size
    } catch {
      return false
    }
  }

  // ── Cloud folder management ────────────────────────────────────────────────

  async _ensureCloudFolder(relFolderPath) {
    if (!relFolderPath || relFolderPath === '.' || relFolderPath === '') return null
    if (this.folderIdCache[relFolderPath]) return this.folderIdCache[relFolderPath]

    const parts = relFolderPath.split(/[/\\]/).filter(Boolean)
    let parentId = null

    for (let i = 0; i < parts.length; i++) {
      const partial   = parts.slice(0, i + 1).join('/')
      const cached    = this.folderIdCache[partial]
      if (cached) { parentId = cached; continue }

      const folder = await this._createCloudFolder(parts[i], parentId)
      this.folderIdCache[partial] = folder.id
      parentId = folder.id
    }

    return parentId
  }

  async _createCloudFolder(name, parentId) {
    try {
      const { data } = await axios.post(
        `${this.apiBase}/api/folders`,
        { name, parentId: parentId || null },
        { headers: this._authHeaders(), timeout: 30000 }
      )
      return data
    } catch (err) {
      // 409 Conflict = folder already exists; fetch its id from the error or re-list
      if (err.response?.status === 409 || err.response?.data?.message?.toLowerCase().includes('already')) {
        // Best-effort: return a fake object so we don't block; the upload will go to root
        return { id: null }
      }
      throw err
    }
  }

  // ── File upload ────────────────────────────────────────────────────────────

  async _uploadFile(filePath) {
    if (!fs.existsSync(filePath)) return

    const rel      = this._rel(filePath)
    const fileName = path.basename(filePath)
    const dirRel   = path.dirname(rel)

    this._status('syncing', `Uploading ${fileName}…`)

    // Resolve cloud folder
    const folderId = await this._ensureCloudFolder(dirRel === '.' ? '' : dirRel)

    const form = new FormData()
    form.append('file', fs.createReadStream(filePath), { filename: fileName })
    if (folderId)  form.append('folderId', String(folderId))
    if (rel)       form.append('relativePath', rel.replace(/\\/g, '/'))

    await axios.post(`${this.apiBase}/api/files/upload`, form, {
      headers: { ...form.getHeaders(), ...this._authHeaders() },
      timeout: 0,
      maxContentLength: Infinity,
      maxBodyLength: Infinity,
    })

    // Update sync state
    const stat = fs.statSync(filePath)
    this.syncState[rel] = { mtime: stat.mtimeMs, size: stat.size }
    this._saveState()

    this.stats.synced++
    this.stats.lastSync = new Date().toISOString()
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  _rel(filePath) {
    return path.relative(this.syncFolder, filePath)
  }

  _authHeaders() {
    return {
      'Authorization': `Bearer ${this.token}`,
      'X-Device-ID':   this.deviceId,
    }
  }

  _status(type, message) {
    this.onStatus({ type, message, stats: this.getStats() })
  }

  _loadState() {
    try { return JSON.parse(fs.readFileSync(this.stateFile, 'utf-8')) } catch { return {} }
  }

  _saveState() {
    try {
      fs.mkdirSync(path.dirname(this.stateFile), { recursive: true })
      fs.writeFileSync(this.stateFile, JSON.stringify(this.syncState, null, 2))
    } catch {}
  }
}

module.exports = SyncEngine

File: desktop-java/src/main/java/com/magizhchi/box/desktop/service/SyncEngine.java
package com.magizhchi.box.desktop.service;

import com.fasterxml.jackson.databind.JsonNode;

import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;

import static java.nio.file.StandardWatchEventKinds.*;

/**
 * Watches a local folder and mirrors new / modified files to the cloud.
 *
 * – Uses Java NIO WatchService (recursive — registers every subdirectory).
 * – Debounces rapid events (500 ms) to avoid duplicate uploads on save.
 * – Change detection: tracks { lastModified, size } so unchanged files are
 *   skipped on restart.
 * – Upload queue: max 2 concurrent uploads via a fixed thread pool.
 */
public class SyncEngine {

    public enum StatusType { IDLE, SYNCING, PAUSED, ERROR }

    public record SyncStatus(StatusType type, String message, int synced, int pending, int errors) {}

    private final ApiService api;
    private final Path syncRoot;

    // Cloud folder cache:  relative POSIX path  →  cloud folder id (String)
    private final Map<String, String> folderIdCache = new ConcurrentHashMap<>();

    // Change-detection:  relative POSIX path  →  "mtime:size"
    private final Map<String, String> syncState = new ConcurrentHashMap<>();

    // Debounce timers keyed by absolute path
    private final Map<Path, ScheduledFuture<?>> debounce = new ConcurrentHashMap<>();

    // Upload queue
    private final ExecutorService uploadPool = Executors.newFixedThreadPool(2);

    // WatchService
    private final WatchService watcher;
    private final Map<WatchKey, Path> watchKeys = new ConcurrentHashMap<>();
    private Thread watchThread;

    // Scheduling for debounce
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

    // Stats
    private final Object statsLock = new Object();
    private int synced  = 0;
    private int pending = 0;
    private int errors  = 0;

    // Paused flag
    private volatile boolean paused = false;

    // Status callback (always called on a non-UI thread — caller must Platform.runLater)
    private final Consumer<SyncStatus> onStatus;

    // Log callback — list of recent activity strings
    private final Consumer<String> onLog;

    private static final DateTimeFormatter TIME_FMT = DateTimeFormatter.ofPattern("HH:mm:ss");

    public SyncEngine(ApiService api, Path syncRoot,
                      Consumer<SyncStatus> onStatus, Consumer<String> onLog) throws IOException {
        this.api       = api;
        this.syncRoot  = syncRoot;
        this.onStatus  = onStatus;
        this.onLog     = onLog;
        this.watcher   = FileSystems.getDefault().newWatchService();
    }

    // ── Public API ─────────────────────────────────────────────────────────────

    public void start() {
        notifyStatus(StatusType.SYNCING, "Starting initial scan…");

        // Initial scan in upload pool (background)
        uploadPool.submit(() -> {
            try {
                initialScan(syncRoot.toFile());
            } catch (Exception e) {
                notifyStatus(StatusType.ERROR, "Scan failed: " + e.getMessage());
            }
        });

        // Start WatchService loop
        watchThread = new Thread(this::watchLoop, "sync-watcher");
        watchThread.setDaemon(true);
        watchThread.start();

        // Register root (and subdirs)
        try {
            registerAll(syncRoot);
        } catch (IOException e) {
            notifyStatus(StatusType.ERROR, "Watch failed: " + e.getMessage());
        }
    }

    public void stop() {
        try { watcher.close(); } catch (IOException ignored) {}
        uploadPool.shutdownNow();
        scheduler.shutdownNow();
        if (watchThread != null) watchThread.interrupt();
    }

    public void pause() {
        paused = true;
        notifyStatus(StatusType.PAUSED, "Sync paused");
    }

    public void resume() {
        paused = false;
        notifyStatus(StatusType.IDLE, "Resuming…");
    }

    // ── File system watching ───────────────────────────────────────────────────

    private void registerAll(Path root) throws IOException {
        Files.walkFileTree(root, new SimpleFileVisitor<>() {
            @Override
            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
                register(dir);
                return FileVisitResult.CONTINUE;
            }
        });
    }

    private void register(Path dir) throws IOException {
        WatchKey key = dir.register(watcher, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
        watchKeys.put(key, dir);
    }

    private void watchLoop() {
        while (!Thread.currentThread().isInterrupted()) {
            WatchKey key;
            try {
                key = watcher.take();
            } catch (InterruptedException | ClosedWatchServiceException e) {
                break;
            }

            Path dir = watchKeys.get(key);
            if (dir == null) { key.reset(); continue; }

            for (WatchEvent<?> event : key.pollEvents()) {
                WatchEvent.Kind<?> kind = event.kind();
                if (kind == OVERFLOW) continue;

                @SuppressWarnings("unchecked")
                Path name  = ((WatchEvent<Path>) event).context();
                Path full  = dir.resolve(name);

                if (kind == ENTRY_CREATE) {
                    if (Files.isDirectory(full)) {
                        // Register new subdirectory
                        try { registerAll(full); } catch (IOException ignored) {}
                    } else {
                        debounce(full, () -> enqueueUpload(full));
                    }
                } else if (kind == ENTRY_MODIFY) {
                    if (!Files.isDirectory(full)) {
                        debounce(full, () -> enqueueUpload(full));
                    }
                }
                // ENTRY_DELETE — not auto-deleting from cloud (safety)
            }

            key.reset();
        }
    }

    private void debounce(Path path, Runnable action) {
        ScheduledFuture<?> existing = debounce.put(path,
                scheduler.schedule(() -> {
                    debounce.remove(path);
                    action.run();
                }, 800, TimeUnit.MILLISECONDS)
        );
        if (existing != null) existing.cancel(false);
    }

    // ── Initial scan ───────────────────────────────────────────────────────────

    private void initialScan(File dir) {
        File[] entries = dir.listFiles();
        if (entries == null) return;
        for (File f : entries) {
            String name = f.getName();
            if (name.startsWith(".") || name.equals("desktop.ini") || name.equals("Thumbs.db")) continue;
            if (f.isDirectory()) {
                initialScan(f);
            } else {
                if (hasChanged(f)) {
                    enqueueUpload(f.toPath());
                }
            }
        }
    }

    private boolean hasChanged(File file) {
        String rel  = relPosix(file.toPath());
        String prev = syncState.get(rel);
        String curr = file.lastModified() + ":" + file.length();
        return !curr.equals(prev);
    }

    // ── Upload ─────────────────────────────────────────────────────────────────

    private void enqueueUpload(Path filePath) {
        if (paused) return;
        if (!Files.exists(filePath) || Files.isDirectory(filePath)) return;
        synchronized (statsLock) { pending++; }
        notifyStatus(StatusType.SYNCING, "Queued: " + filePath.getFileName());

        uploadPool.submit(() -> {
            try {
                uploadFile(filePath);
            } catch (Exception e) {
                synchronized (statsLock) { errors++; pending = Math.max(0, pending - 1); }
                notifyStatus(StatusType.ERROR, "Failed: " + filePath.getFileName() + " — " + e.getMessage());
                log("✗ " + filePath.getFileName() + " — " + e.getMessage());
            }
        });
    }

    private void uploadFile(Path filePath) throws IOException {
        if (!Files.exists(filePath)) return;

        File  file    = filePath.toFile();
        String rel    = relPosix(filePath);
        String dirRel = posixParent(rel);

        notifyStatus(StatusType.SYNCING, "Uploading " + file.getName() + "…");

        // Ensure the parent cloud folder exists
        String folderId = dirRel.isEmpty() ? null : ensureCloudFolder(dirRel);

        api.uploadFile(file, folderId, rel);

        // Update change-detection state
        syncState.put(rel, file.lastModified() + ":" + file.length());

        synchronized (statsLock) {
            synced++;
            pending = Math.max(0, pending - 1);
        }

        String time = LocalTime.now().format(TIME_FMT);
        log("[" + time + "] ✓ " + rel);

        notifyStatus(pending == 0 ? StatusType.IDLE : StatusType.SYNCING,
                pending == 0
                        ? "Up to date · " + synced + " file(s) synced"
                        : "Uploading… (" + pending + " remaining)");
    }

    // ── Cloud folder management ────────────────────────────────────────────────

    /** Returns the cloud folder ID for the given POSIX-style relative path, creating it if needed. */
    private String ensureCloudFolder(String relFolderPath) throws IOException {
        if (relFolderPath == null || relFolderPath.isEmpty()) return null;

        String cached = folderIdCache.get(relFolderPath);
        if (cached != null) return cached;

        String[] parts    = relFolderPath.split("/");
        String   parentId = null;

        for (int i = 0; i < parts.length; i++) {
            String partial = String.join("/", Arrays.copyOfRange(parts, 0, i + 1));
            String hit     = folderIdCache.get(partial);
            if (hit != null) { parentId = hit; continue; }

            String folderId = createCloudFolder(parts[i], parentId);
            folderIdCache.put(partial, folderId);
            parentId = folderId;
        }

        return parentId;
    }

    private String createCloudFolder(String name, String parentId) throws IOException {
        try {
            JsonNode resp = api.createFolder(name, parentId);
            return resp.get("id").asText();
        } catch (IOException e) {
            // Folder may already exist — try to tolerate 409/conflict
            if (e.getMessage() != null && (e.getMessage().contains("409") ||
                    e.getMessage().toLowerCase().contains("already"))) {
                return null; // fall back to root
            }
            throw e;
        }
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    /** Relative path from sync root, using forward slashes. */
    private String relPosix(Path filePath) {
        return syncRoot.relativize(filePath).toString().replace('\\', '/');
    }

    /** Parent directory of a POSIX relative path, or "" for root. */
    private String posixParent(String posixPath) {
        int idx = posixPath.lastIndexOf('/');
        return idx < 0 ? "" : posixPath.substring(0, idx);
    }

    private void notifyStatus(StatusType type, String message) {
        int s, p, e;
        synchronized (statsLock) { s = synced; p = pending; e = errors; }
        onStatus.accept(new SyncStatus(type, message, s, p, e));
    }

    private void log(String line) {
        onLog.accept(line);
    }
}
